nengo/tests/test_network.py

```diff
@@ -67,28 +67,52 @@ def test_nested_context():
 
 
 def test_context_errors():
+    """Most of these errors are in ``Network.add``; one is in ``Network.__exit__``"""
+
     def add_something():
         nengo.Ensemble(1, dimensions=1)
 
     # Error if adding before Network creation
-    with pytest.raises(RuntimeError):
+    with pytest.raises(NetworkContextError, match="must either be created inside a"):
         add_something()
 
     model = nengo.Network()
     # Error if adding before a `with network` block
-    with pytest.raises(RuntimeError):
+    with pytest.raises(NetworkContextError, match="must either be created inside a"):
         add_something()
 
     # Error if adding after a `with network` block
     with model:
         add_something()
-    with pytest.raises(RuntimeError):
+    with pytest.raises(NetworkContextError, match="must either be created inside a"):
         add_something()
 
+    # Error if adding a bad object type
+    with model:
+        with pytest.raises(NetworkContextError, match="Objects of type.*added to net"):
+            model.add(object())
+
     # Okay if add_to_container=False
     nengo.Ensemble(1, dimensions=1, add_to_container=False)
     nengo.Node(output=[0], add_to_container=False)
 
+    # Errors if context is in a bad state (one in `Network.__exit__`)
+    with pytest.raises(NetworkContextError, match="bad state; was expecting current"):
+        with nengo.Network():
+            nengo.Network.context.append("bad")
```

Comment:
I would have thought that messing with the global contexts like this would have bad effects in other tests (since we're never restoring it to a good state). But none of the other tests are failing. Is the context actually being restored in some way, or is Nengo in general just insensitive to these contexts being in a bad state?

Reply:
That's a good point. My guess is we're just insensitive to it. For example, when you enter a network `with` block, both that network and its config get added to their respective lists. Then when we're using them, we're always looking at the last one in each list, so it doesn't really matter if one is longer than the other. Then when we exit, we remove from both.

We should probably use `monkeypatch` for this, though, just to be safe.